### DISCUSSION TITLE
AI : Add support for NN inference

### DISCUSSION THREAD

**Discussion by WilliamTambellini** [o3de/o3de#5251] [General]
**Is your feature request related to a problem? Please describe.**
There is apparently today no way to run any NN inference using the core engine.
There is also actually no easy way to infer a linear layer (Output=Input*Weights+Bias) 
because there is apparently no general MxN Matrix class:
https://github.com/o3de/o3de/tree/development/Code/Framework/AzCore/AzCore/Math

**Describe the solution you'd like**
At least 3 solutions :

A- a Tensor class to do the regular Tensor algebra. 
Some references:
the eigen Tensor class, at the core of TensorFlow:
https://eigen.tuxfamily.org/dox/unsupported/eigen_tensors.html
the ATen Tensor class, the core of pytorch:
https://pytorch.org/cppdocs/

B- Adding a minimalist (header only) NN engine in order to infer some well known layers (linear, activation, reduction, ...). 
Examples:
https://github.com/yixuan/MiniDNN
https://github.com/Dobiasd/frugally-deep

C- Linking with a full featured generic NN library capable to infer a NN entirely defined in a file/stream or created on the fly. 
Examples:
https://www.mlpack.org
https://onnx.ai
https://github.com/microsoft/onnxruntime

**Describe alternatives you've considered**
None

**Additional context**
In 2017, Unity officially released 'Barracuda', the 'Unity-Inference-Engine', to at least run inference on a NN in the official Unity :
https://docs.unity3d.com/Packages/com.unity.barracuda@1.0/manual/index.html
https://github.com/Unity-Technologies/ml-agents/blob/main/docs/Unity-Inference-Engine.md

**Comment by bhack** [comment]
Some other interesting runtimes:
https://github.com/tensorflow/runtime
https://github.com/google/iree/
